### PR TITLE
Refactor static widgets

### LINKED ISSUE #800: Alternate embed manager supporting custom views in tagged divs
At the moment, when embedding widgets in a static web pages, all the widgets of the notebook are displayed in a single column. (The script tag containing the widget manager state is replaced with the rendered version of the widgets).

Instead, what we could do is allow the document author to specify which views he want with specific tags:

``` html
<script tag linking to embed manager code>
<script tag with manager state>
[later in the page]
<div class="widget_view" model_id="AF13B09">
[...]
<div class="widget_view" model_id="23AA2B8">
```

### LINKED ISSUE #907: Embedding mechanism
**Current Mechanism used on http://jupyter.org/widgets**

1. Embed manager script tag from cdn
2. Script tag with `type=application/vnd.jupyter-embedded-widgets` containing the widget manager state.

    The format for this state is the one described here:

    https://github.com/ipython/ipywidgets/blob/master/ipywidgets/state.schema.json

    before we wrote the spec, so without the version information.

    Besides, each widget has a `views` attribute of type array. If it is not empty, the widget is a *top-level*  widget.

    This script tag is picked up by the embed manager and replaced by a rendered version of all the top-level widgets in a row.

**Proposed mechanism for web pages**

1. Embed manager script tag from cdn
2. Script tag with `type=application/vnd.jupyter.widget-state+json` containing the widget manager state, following precisely the schema for the manager state

    https://github.com/ipython/ipywidgets/blob/master/ipywidgets/state.schema.json

3. Script tags with `type=application/vnd.jupyter.widget-view+json` containing

    ```json
    {
        "model_id": {
            "type": "string",
            "description": "Model id to be displayed inplace"
        }
    }
    ```

**Proposed mechanism for nbviewer**

1. nbviewer loads the embed manager JavaScript, either from cdn or bundles is

2. Each output area containing widgets gets its `type=application/vnd.jupyter.widget-view+json` tag in the html mime bundle.

3. Global widget state following the state schema is saved in the notebook metadata.

    The html nbconversion picks up that metadata and incorporates a `type=application/vnd.jupyter.widget-state+json`.

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Implementation of the #907 proposal for the embedding mechanism and json specification of widget state....

Closes #907
Closes #800

Here is the example of HTML that it produces

```json
<script src="https://unpkg.com/jupyter-js-widgets@~2.0.7/dist/embed.js"></script>
<script type="application/vnd.jupyter.widget-state+json">
{
    "version_major": 1,
    "version_minor": 0,
    "state": {
        "350f6bc9674a4caeafe5cca7a8496ce8": {
            "model_name": "LayoutModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {}
        },
        "67693fb358624b71b199ee0ffae88991": {
            "model_name": "IntSliderModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {
                "value": 7,
                "readout_format": "i",
                "continuous_update": false,
                "description": "Test:",
                "slider_color": "white",
                "max": 10,
                "layout": "IPY_MODEL_350f6bc9674a4caeafe5cca7a8496ce8"
            }
        },
        "5733b0f94b6a4bb8ade382ecb3979565": {
            "model_name": "LayoutModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {}
        },
        "d946e9da816e4f99b551f3bca3396882": {
            "model_name": "FloatSliderModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {
                "step": 0.1,
                "value": 7.5,
                "readout_format": ".1f",
                "continuous_update": false,
                "description": "Test:",
                "slider_color": "white",
                "max": 10,
                "layout": "IPY_MODEL_5733b0f94b6a4bb8ade382ecb3979565"
            }
        },
        "12774db31fa54e09a9d623b4e27e4bcd": {
            "model_name": "LayoutModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {}
        },
        "23e693118ae34e25acd0f8dc831fc779": {
            "model_name": "FloatSliderModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {
                "step": 0.1,
                "orientation": "vertical",
                "value": 7.5,
                "readout_format": ".1f",
                "continuous_update": false,
                "description": "Test:",
                "slider_color": "white",
                "max": 10,
                "layout": "IPY_MODEL_12774db31fa54e09a9d623b4e27e4bcd"
            }
        },
        "9724b40f40f34a1cae9f9c7c2171e07b": {
            "model_name": "LayoutModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {}
        },
        "ee0e018acd90450ea5a6e4dd9c082bba": {
            "model_name": "IntSliderModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {
                "_range": true,
                "value": [
                    5,
                    7
                ],
                "layout": "IPY_MODEL_9724b40f40f34a1cae9f9c7c2171e07b",
                "description": "Test:",
                "slider_color": "white",
                "max": 10,
                "continuous_update": false
            }
        },
        "d0f62afb967344ffa76189fd3d1c6f0f": {
            "model_name": "LayoutModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {}
        },
        "3f49dcc9b7a746d4ad432b3a7f5838f7": {
            "model_name": "FloatSliderModel",
            "model_module": "jupyter-js-widgets",
            "model_module_version": "*",
            "state": {
                "step": 0.1,
                "_range": true,
                "value": [
                    5,
                    7.5
                ],
                "layout": "IPY_MODEL_d0f62afb967344ffa76189fd3d1c6f0f",
                "description": "Test:",
                "slider_color": "white",
                "max": 10,
                "continuous_update": false
            }
        }
    }
}
</script>
<script type="application/vnd.jupyter.widget-view+json">
{
    "model_id": "67693fb358624b71b199ee0ffae88991"
}
</script>
<script type="application/vnd.jupyter.widget-view+json">
{
    "model_id": "d946e9da816e4f99b551f3bca3396882"
}
</script>
<script type="application/vnd.jupyter.widget-view+json">
{
    "model_id": "23e693118ae34e25acd0f8dc831fc779"
}
</script>
<script type="application/vnd.jupyter.widget-view+json">
{
    "model_id": "ee0e018acd90450ea5a6e4dd9c082bba"
}
</script>
<script type="application/vnd.jupyter.widget-view+json">
{
    "model_id": "3f49dcc9b7a746d4ad432b3a7f5838f7"
}
</script>
```

### CODE REVIEW

File: jupyter-js-widgets/examples/web4/index.html [hunk 1]
```diff
@@ -13,92 +13,92 @@
...
+                        "layout": "IPY_MODEL_4e6ee45ef7c64f0f953c79150d4ae0ca"
+                    }
+                }
             }
         }
         </script>
+        <script type="application/vnd.jupyter.widget-view+json">
+        {
+            "model_id": "4c5a49e52ae54ad5bdbe973353a58931"
+        }
```
Comment (SylvainCorlay) [inline jupyter-js-widgets/examples/web4/index.html]: Thanks. Looking into it.

**Comment by SylvainCorlay** [comment]
This is ready for review. I would like to get this in quickly and iterate.

It depends on @gnestor 's merged PR to nbformat (https://github.com/jupyter/nbformat/pull/69)

As soon as a patch version to nbformat is released, we can add the dependency to the latest version.

**Comment by jasongrout** [comment]
> As soon as a patch version to nbformat is released, we can add the dependency to the latest version.

I'd like to keep iterating on betas. Do we need to point to a new version of nbformat for master to work? If so, let's hold off on merging so we can keep releasing beta versions that work.

**Comment by SylvainCorlay** [comment]
> I'd like to keep iterating on betas. Do we need to point to a new version of nbformat for master to work? If so, let's hold off on merging so we can keep releasing beta versions that work.

@minrk promised a version of nbformat by friday. It currently works but pops up a warning when saving notebooks with widgets.

**Comment by jasongrout** [comment]
> It depends on @gnestor 's merged PR to nbformat (jupyter/nbformat#69)

Does the running code depend on his merged PR, or is that just a spec change which can be done in parallel?

**Comment by SylvainCorlay** [comment]
The running code saves `???+json` things to the notebook which fails to validate with nbformat because of the bug fixed by grant's pr.

**Pull request merged**